### PR TITLE
Expose IGridSampleLayer

### LINKED ISSUE #2202: Expose IGridSampleLayer in dynamo.conversion.impl

### PR DESCRIPTION
**Pull request by apbose**
Fixes #2202

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 1]
```diff
@@ -245,7 +245,40 @@ def aten_ops_fmod(
     return impl.elementwise.fmod(ctx, target, SourceIR.ATEN, name, args[0], args[1])
 
 
-@dynamo_tensorrt_converter(torch.ops.aten.relu.default)  # type: ignore[misc]
+@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler)
+@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler_2d.out)
+@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler_2d_backward.out)
+@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler_3d.out)
+@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler_3d_backward.out)
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:252]: Add `# type: ignore[misc]` to these

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 2]
```diff
@@ -245,7 +245,40 @@ def aten_ops_fmod(
     return impl.elementwise.fmod(ctx, target, SourceIR.ATEN, name, args[0], args[1])
 
 
-@dynamo_tensorrt_converter(torch.ops.aten.relu.default)  # type: ignore[misc]
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:248]: Undo this change

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 1 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:252]: Remove `_backward` implementations. Consider changing the decorator stack to:
```python
@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler.default)  # type: ignore[misc]
@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler_2d.default)  # type: ignore[misc]
@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler_3d.default)  # type: ignore[misc]
```

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 3]
```diff
@@ -245,7 +245,40 @@ def aten_ops_fmod(
...
+        ctx, 
+        target, 
+        SourceIR.ATEN, 
+        name, 
+        input=args[0], 
+        grid=args[1], 
+        interpolation_mode=args[2], 
+        padding_mode=args[3], 
+        align_corners=args_bounds_check(args, 4, True),
+        output_mask=args_bounds_check(args, 5, None),
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:276]: This can be removed once the `out` variants are removed from the decorator list

File: py/torch_tensorrt/dynamo/conversion/converter_utils.py [hunk 4]
```diff
@@ -24,6 +24,36 @@
...
+        self.sample_mode = None
+
+    def __call__(self, sample_int):
+        if sample_int == 0:
+            self.sample_mode = trt.SampleMode.FILL
+        elif sample_int == 1:
+            self.sample_mode = trt.SampleMode.CLAMP
+        elif sample_int == 2:
+            self.sample_mode = trt.SampleMode.REFLECT
+        return self.sample_mode
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/converter_utils.py:54]: I think this could just go in the `grid.py` file, since it is very specific to the grid operations.

File: py/torch_tensorrt/dynamo/conversion/impl/grid.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    grid: TRTTensor,
+    interpolation_mode: int,
+    padding_mode: int,
+    align_corners: bool,
+    output_mask: Optional[Sequence[bool]] = None,
+) -> TRTTensor:
+    grid_layer = ctx.net.add_grid_sample(input, grid)
+    interpolation_mode_trt = GridSamplerInterpolation()
+    grid_layer.interpolation_mode = interpolation_mode_trt(interpolation_mode)
+    sample_mode_trt = GridSamplerSampling()
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/grid.py:32]: Consider using a dictionary for this - could be `grid_layer.interpolation_mode = Dictionary.get(interpolation_mode, None)`

File: py/torch_tensorrt/dynamo/conversion/impl/grid.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    set_layer_name(grid_layer, target, name + "_grid_layer", source_ir)
+    if output_mask is None:
+        return grid_layer.get_output(0)
+    else:
+        if output_mask[0] and output_mask[1]:
+            return (grid_layer.get_output(0), None)
+        elif output_mask[0]:
+            return grid_layer.get_output(0)
+        else:
+            return None
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/impl/grid.py:44]: Bring up one level and make all of these `elif` statements

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 3 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:276]: Do you mean the` output_mask` argument? As far as I understood, the `.backward` would have the `output_mask` argument.

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 2 above)
Comment (apbose) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:248]: Oh I think this got deleted in the formatting change, it is present below

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 1 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:252]: Regarding the `# type: ignore[misc]`, generally `mypy` linting will indicate an error for untyped decorators. Our `dynamo_tensorrt_converter` is not untyped, but for some reason the error persists. For now, we just keep the `ignore` on all decorators in this file.

On the second issue, I think you can remove the `.out` variant, but change the `@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler.default)` back to just `@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler)`. That should address the error

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 3 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:276]: Yes, I meant the `output_mask` argument. I don't think we should ever encounter an argument of the `backward` variety since the backend is strictly forward/inference-only

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py (hunk 2 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:248]: I just intended to have the `ignore` be kept

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 7]
```diff
@@ -245,6 +245,33 @@ def aten_ops_fmod(
     return impl.elementwise.fmod(ctx, target, SourceIR.ATEN, name, args[0], args[1])
 
 
+@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler)  # type: ignore[misc]
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:248]: It seems that plain `grid_sampler_2d` and `grid_sampler_3d` are also valid schemas - could those be added here as well?

File: tests/py/dynamo/conversion/test_grid_aten.py
Comment (gs-olive) [inline tests/py/dynamo/conversion/test_grid_aten.py:1]: Could tests also be added for 2D and 3D grid sample. Looks good to me other than that

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 8]
```diff
@@ -245,6 +245,35 @@ def aten_ops_fmod(
...
+    return impl.grid.grid(
+        ctx,
+        target,
+        SourceIR.ATEN,
+        name,
+        input=args[0],
+        grid=args[1],
+        interpolation_mode=args[2],
+        padding_mode=args[3],
+        align_corners=args_bounds_check(args, 4, True),
```
[the remark this answers is not in the document]
Comment (apbose) [reply py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:273]: Thanks for pointing it out. I think earlier I had looked at the `cudnn_grid_sampler` which made me think it is optional. I will change it.

**Review by gs-olive — COMMENTED** [review]
May need a quick linter run for `grid.py`, and just needs copies of the existing tests for `2D` and `3D` - then it should be good to go!

**Comment by apbose** [comment]
Hi @gs-olive, is there anything else to be addressed in this PR? I do not see anything, but I could have missed in case any,

**Review by narendasan — APPROVED** [review]
LGTM

**Comment by apbose** [comment]
Hi @gs-olive regarding adding the 2d and 3d cases, from documentation could you understand which cases will give `grid_sampler_2d` or` grid_sampler_3d`. I tried making the grid last dimension 3 in place of 2, and also tried changing the rank dimension of the input and grid, but they do not seem to be valid inputs to the IGrid Layer

**Comment by gs-olive** [comment]
@apbose - I believe in the test case, you would have to change `torch.ops.aten.grid_sampler` to `torch.ops.aten.grid_sampler_2d` and `torch.ops.aten.grid_sampler_3d`, so as to directly invoke those functions. Additionally, [this documentation](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html) from `nn.grid_sample` could be helpful. If the tests do not seem to be working even with these modifications, it is okay to remove the decorators for the 2D and 3D cases.

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 9]
```diff
@@ -245,6 +245,36 @@ def aten_ops_fmod(
     return impl.elementwise.fmod(ctx, target, SourceIR.ATEN, name, args[0], args[1])
 
 
+@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler)  # type: ignore[misc]
+@dynamo_tensorrt_converter(torch.ops.aten.grid_sampler_2d)  # type: ignore[misc]
+# commented this for now, see py/dynamo/conversion/tests/test_grid_aten. Should this be removed altogether?
+# @dynamo_tensorrt_converter(torch.ops.aten.grid_sampler_3d)  # type: ignore[misc]
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:251]: If the tests are not working, it is okay to remove the 3d case altogether

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 10]
```diff
@@ -245,6 +245,36 @@ def aten_ops_fmod(
...
+@enforce_tensor_types(
+    {
+        0: (TRTTensor,),
+        1: (TRTTensor,),
+    }
+)  # type: ignore[misc]
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:257]: The ` # type: ignore[misc]` can be removed from the decorators

**Comment by apbose** [comment]
Closed in favor of #2457

**Review by gs-olive — APPROVED** [review]
Looks good to me!
